### PR TITLE
Improve payload cleaning performance

### PR DESCRIPTION
**Pull request by imjoehaines**
## Goal

This PR improves performance when sending notifications by only recursively iterating over the report once, rather than doing it several times

We need to iterate through the full payload because we need to filter potentially sensitive data from `metaData`, fixup any encoding issues, fix recursive objects and objects that raise when being stringified

Previously we would iterate through the report `metaData` (and breadcrumb metadata) with filtering enabled, iterate through the rest of the report without filtering it (as we don't want to filter outside of `metaData`) and then iterate through the entire payload (which includes the report) in order to fix encoding/recursion/exceptions

Now we iterate over the full payload in one go, which saves a lot of CPU time

### Changed

The main `Bugsnag#notify` method now takes care of filtering as well as fixing up the payload. It knows about which parts of the payload it should apply filtering to based on the new `SCOPES_TO_FILTER`. This is the same way that the [JavaScript notifier applies filtering](https://github.com/bugsnag/safe-json-stringify#usage), for the same performance reasons

To do this the `Bugsnag::Cleaner` now takes a `scopes_to_filter` parameter on initialisation, which is uses to check if a scope needs to be filtered

This also means that the `Helper` class doesn't apply cleaning itself, but now only trims the payload if it's too long

### Linked issues

Related to #481 

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [x] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: lib/bugsnag.rb [hunk 1]
```diff
@@ -274,7 +298,28 @@ def check_endpoint_setup
...
+    ##
+    # Convert the Report object to JSON
+    #
+    # We ensure the report is safe to send by removing recursion, fixing
+    # encoding errors and redacting metadata according to "meta_data_filters"
+    #
+    # @param report [Report]
+    # @return string
+    def report_to_json(report)
+      cleaner = Cleaner.new(
```
[the remark this answers is not in the document]
Comment (imjoehaines) [reply lib/bugsnag.rb:311]: Yeah; if it's cached then the `configuration.meta_data_filters` can get outdated if they're changed after the first time it's created

This could be solved by giving `Cleaner` access to the `Configuration` directly, but I'm not sure that's better

File: lib/bugsnag/cleaner.rb [hunk 2]
```diff
@@ -88,7 +94,7 @@ def clean_string(str)
     end
 
     def self.clean_object_encoding(obj)
-      new(nil).clean_object(obj)
+      new(Set.new, []).clean_object(obj)
```
[the remark this answers is not in the document]
Comment (imjoehaines) [reply lib/bugsnag/cleaner.rb]: Yep, good spot 👍

File: lib/bugsnag.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (imjoehaines) [reply lib/bugsnag.rb:311]: After thinking about it more, it does make more sense to give it the `Configuration` object and construct it once. It then also cleans up the `rack_request` which currently makes its own cleaner which was slightly awkward with the new parameter, but after the change it can get the single shared instance instead

File: lib/bugsnag.rb [hunk 3]
```diff
@@ -212,27 +209,40 @@ def leave_breadcrumb(name, meta_data={}, type=Bugsnag::Breadcrumbs::MANUAL_BREAD
...
+    # Returns the client's Cleaner object, or creates one if not yet created.
+    #
+    # @api private
+    #
+    # @return [Cleaner]
+    def cleaner
+      @cleaner = nil unless defined?(@cleaner)
+      @cleaner || LOCK.synchronize do
+        @cleaner ||= Bugsnag::Cleaner.new(configuration)
       end
```
Comment (kattrali) [inline lib/bugsnag.rb:240]:
```suggestion
      @cleaner ||= LOCK.synchronize do
        Bugsnag::Cleaner.new(configuration)
      end
```

[very tiny nit] You can move the `=` up to make or/assign blocks easier to read.

<details>
<summary>test case</summary>

```rb
class Foo
  LOCK = Mutex.new

  def self.foo
    @foo ||= LOCK.synchronize do
      puts "Assigning value"
      5
    end
  end
end

puts Foo.foo
puts Foo.foo
```
prints:
```
Assigning value
5
5
```

</details>

Comment (imjoehaines) [reply lib/bugsnag.rb:240]: I think the current way is actually required to be thread safe, which doesn't _really_ matter in this case but it's consistent with the `configuration` and `session_tracker` methods

If multiple threads manage to call this at the exact same time, they can both get into the `synchronize` block and set `@cleaner` to a new instance, so I think it's necessary to do this even if it is a bit ugly :p

<details>
<summary>test case</summary>

```ruby
class Foo
  LOCK = Mutex.new

  def self.foo
    @foo ||= LOCK.synchronize do
      puts "Assigning value in #{Thread.current}"
      rand(100)
    end
  end
end

threads = []

10.times do
  threads << Thread.new { puts "[#{Thread.current}] #{Foo.foo}" }
end

threads.each(&:join)
```

This will sometimes set `foo` once, but sometimes multiple threads manage to set it and each sets it to a random number

Whereas this test case will only set `foo` once, even if multiple threads make it into the `synchronize` block:

```ruby
class Foo
  LOCK = Mutex.new

  def self.foo
    @foo || LOCK.synchronize do
      puts "Assigning value in #{Thread.current}"
      @foo ||= rand(100)
    end
  end
end

threads = []

10.times do
  threads << Thread.new { puts "[#{Thread.current}] #{Foo.foo}" }
end

threads.each(&:join)
```
</details>

Comment (kattrali) [reply lib/bugsnag.rb:240]: Aha! I wondered that but wasn't sure how to check it. Thanks!
